### PR TITLE
Don't check against root directory

### PR DESCRIPTION
**Pull request by frostming**
### The issue

#3386 causes some regression issue #4273 

### The fix

This PR reverts some of the changes of #3386, don't check against root directory but fix the `self.name` reference in `create_pipfile()`.

Sorry for the trouble.

### The checklist

* [x] Associated issue
* [x] A news fragment in the `news/` directory to describe this fix with the extension `.bugfix`, `.feature`, `.behavior`, `.doc`. `.vendor`. or `.trivial` (this will appear in the release changelog). Use semantic line breaks and name the file after the issue number or the PR #.

### CODE REVIEW

**Comment by techalchemy** [comment]
The main reason for the change in the first place is due to locating virtual environments & related python paths -- as far as I was aware, this was a substantial cause of bugs and breakages and basically didn't work. The fact that it is breaking workflows is the first I am hearing of it working *at all*.

File: pipenv/project.py [hunk 1]
```diff
@@ -685,8 +685,8 @@ def create_pipfile(self, python=None):
         from .vendor.pip_shims.shims import (
             ConfigOptionParser, make_option_group, index_group
         )
-
-        config_parser = ConfigOptionParser(name=self.name)
+        # Inherit the pip's index configuration of install command.
+        config_parser = ConfigOptionParser(name="install")
```
Comment (jayaddison) [inline pipenv/project.py:689]: Is this change intended, @frostming?  It doesn't seem related to the previous root-path handling modification (#3386)

Comment (jayaddison) [reply pipenv/project.py:689]: (to clarify, on taking a second look - this change does appear to affect a similar section of code to the previous modifications, but it's not immediately clear whether it's a direct revert of them; it may be the correct way to go, but I figured it'd be worth checking just in case)

Comment (frostming) [reply pipenv/project.py:689]: The original issue is that when under the root directory, `ConfigOptionParser` will complain that `sel.name` is empty. The fix in #3386 prevented it fron happening at early stage, which brought a breakage in many users' docker building setup.

So I revert the early checking and instead make `ConfigOptionaParser` happy by passing a meaningful section name to it.
Because giving an arbitrary name doesn't make any sense, which i didn't realize before. This should be the correct path to resolving the issue.

Comment (techalchemy) [inline pipenv/project.py:689]: Arguably, we should change this to `pip_shims.shims.InstallCommand()` and grab the relevant option groups from there. Trying to construct this ourselves out of option groups is foolish and I don't think it actually works correctly.

```suggestion
        command = InstallCommand()
        extra_indexes = command.cmd_opts.get_option("--extra-index-url").default
```

```python
>>> from pip_shims.shims import InstallCommand
>>> cmd = InstallCommand()
>>> cmd.cmd_opts.get_option("--extra-index-url").default
[]
```

File: pipenv/project.py [hunk 2]
```diff
@@ -685,8 +685,8 @@ def create_pipfile(self, python=None):
         from .vendor.pip_shims.shims import (
             ConfigOptionParser, make_option_group, index_group
```
Comment (techalchemy) [inline pipenv/project.py:686]:
```suggestion
            InstallCommand
```

**Comment by techalchemy** [comment]
I think the changes make sense overall -- my only question is -- wasn't one of our main issues regarding the `which` functionality that is partly derived from project locations? If that is completely resolved (which it should be, I would think) then we shouldn't have any issues allowing pipenv usage from `/`

File: pipenv/project.py [hunk 3]
```diff
@@ -685,8 +685,8 @@ def create_pipfile(self, python=None):
...
         )
-
-        config_parser = ConfigOptionParser(name=self.name)
+        # Inherit the pip's index configuration of install command.
+        config_parser = ConfigOptionParser(name="install")
         config_parser.add_option_group(make_option_group(index_group, config_parser))
```
Comment (techalchemy) [inline pipenv/project.py:690]:
```suggestion
```

File: pipenv/project.py [hunk 4]
```diff
@@ -685,8 +685,8 @@ def create_pipfile(self, python=None):
...
         install = config_parser.option_groups[0]
```
Comment (techalchemy) [inline pipenv/project.py:691]:
```suggestion
        install = config_parser.option_groups[0]
        indexes = command.cmd_opts.get_option("--extra-index-url").default
```

and drop the `indexes` defined in the line below

**Comment by frostming** [comment]
> One small update to how we are determining extra index url defaults -- our syntax here is kind of outdated

Agree, do as suggested

>my only question is -- wasn't one of our main issues regarding the which functionality that is partly derived from project locations? If that is completely resolved (which it should be, I would think) then we shouldn't have any issues allowing pipenv usage from /

Yes, when under the root directory `self.name` is just an empty string, which can work in subsequent location calculations.
https://github.com/pypa/pipenv/blob/3f4abadac255d527f94c8fa60590e1472acd814b/pipenv/project.py#L220-L224

File: pipenv/core.py [hunk 5]
```diff
@@ -126,16 +126,15 @@ def do_clear():
...
-        )
     except OSError as e:
         # Ignore FileNotFoundError. This is needed for Python 2.7.
         import errno
 
         if e.errno == errno.ENOENT:
             pass
         raise
+    # Other processes may be writing into this directory simultaneously.
+    vistir.path.rmtree(locations.USER_CACHE_DIR, ignore_errors=True)
```
Comment (techalchemy) [inline pipenv/core.py:137]: yeah i guess that's fine, or maybe `ignore_errors=environments.PIPENV_IS_CI`

Comment (techalchemy) [reply pipenv/core.py:137]: (but I'm not sure that's necessary, and don't delay merging over it)

Comment (frostming) [reply pipenv/core.py:137]: I didn't know that env var, that looks better.

File: pipenv/project.py (hunk 1 above)
Comment (jayaddison) [reply pipenv/project.py]: Thank you for the explanation!
